File: android/src/main/kotlin/com/snowplowanalytics/snowplow_tracker/readers/entities/MediaPlayerEntityReader.kt
// Copyright (c) 2022-present Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

package com.snowplowanalytics.snowplow_tracker.readers.entities

import com.snowplowanalytics.snowplow.media.entity.MediaPlayerEntity
import com.snowplowanalytics.snowplow.media.entity.MediaType

class MediaPlayerEntityReader(values: Map<String, Any>) {
    private val valuesDefault = values.withDefault { null }

    val currentTime: Double? by valuesDefault
    val duration: Double? by valuesDefault
    val ended: Boolean? by valuesDefault
    val fullscreen: Boolean? by valuesDefault
    val livestream: Boolean? by valuesDefault
    val label: String? by valuesDefault
    val loop: Boolean? by valuesDefault
    val mediaType: String? by valuesDefault
    val muted: Boolean? by valuesDefault
    val paused: Boolean? by valuesDefault
    val pictureInPicture: Boolean? by valuesDefault
    val playerType: String? by valuesDefault
    val playbackRate: Double? by valuesDefault
    val quality: String? by valuesDefault
    val volume: Int? by valuesDefault

    val mediaTypeEnum: MediaType? by lazy {
        when (mediaType) {
            "audio" -> MediaType.Audio
            "video" -> MediaType.Video
            else -> null
        }
    }

    fun toMediaPlayerEntity(): MediaPlayerEntity {
        return MediaPlayerEntity(
            currentTime = currentTime,
            duration = duration,
            ended = ended,
            fullscreen = fullscreen,
            livestream = livestream,
            label = label,
            loop = loop,
            mediaType = mediaTypeEnum,
            muted = muted,
            paused = paused,
            pictureInPicture = pictureInPicture,
            playerType = playerType,
            playbackRate = playbackRate,
            quality = quality,
            volume = volume
        )
    }
}

File: ios/Classes/readers/entities/MediaPlayerEntityReader.swift
// Copyright (c) 2022-present Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

import Foundation
import SnowplowTracker

struct MediaPlayerEntityReader: Decodable {
    let currentTime: Double?
    let duration: Double?
    let ended: Bool?
    let fullscreen: Bool?
    let livestream: Bool?
    let label: String?
    let loop: Bool?
    let mediaType: String?
    let muted: Bool?
    let paused: Bool?
    let pictureInPicture: Bool?
    let playerType: String?
    let playbackRate: Double?
    let quality: String?
    let volume: Int?
    
    var mediaTypeEnum: MediaType? {
        if let mediaType = self.mediaType {
            switch mediaType {
            case "audio":
                return MediaType.audio
            case "video":
                return MediaType.video
            default:
                return nil
            }
        }
        return nil
    }
}

extension MediaPlayerEntityReader {
    func toMediaPlayerEntity() -> MediaPlayerEntity {
        let player = MediaPlayerEntity(
            currentTime: self.currentTime,
            duration: self.duration,
            ended: self.ended,
            fullscreen: self.fullscreen,
            livestream: self.livestream,
            label: self.label,
            loop: self.loop,
            mediaType: self.mediaTypeEnum,
            muted: self.muted,
            paused: self.paused,
            pictureInPicture: self.pictureInPicture,
            playerType: self.playerType,
            playbackRate: self.playbackRate,
            quality: self.quality,
            volume: self.volume
        )
        return player
    }
}
